File: backend/app/routers/bookings.py
"""Booking routes for regular users."""
from datetime import date as date_type

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel

from app.database import get_conn
from app.deps import get_current_user

HALL_OPEN = "06:00"
HALL_CLOSE = "20:00"


def _validate_booking_fields(body: "BookingRequest"):
    if body.date < date_type.today().isoformat():
        raise HTTPException(status_code=400, detail="Cannot book a date in the past")
    if body.start_time < HALL_OPEN or body.start_time >= HALL_CLOSE:
        raise HTTPException(status_code=400, detail=f"Start time must be between {HALL_OPEN} and {HALL_CLOSE}")
    if body.end_time > HALL_CLOSE:
        raise HTTPException(status_code=400, detail=f"End time cannot be after {HALL_CLOSE}")
    if body.end_time <= body.start_time:
        raise HTTPException(status_code=400, detail="End time must be after start time")
    if not (1 <= body.attendees <= 50):
        raise HTTPException(status_code=400, detail="Attendees must be between 1 and 50")


def _check_user_slot_conflict(conn, user_id: int, date: str, start_time: str, end_time: str, exclude_id: int | None = None):
    """Raise 409 if the user already has a pending/approved booking overlapping this slot."""
    query = """
        SELECT id FROM bookings
        WHERE user_id = ? AND date = ? AND status IN ('pending', 'approved')
        AND start_time < ? AND end_time > ?
    """
    params = [user_id, date, end_time, start_time]
    if exclude_id is not None:
        query += " AND id != ?"
        params.append(exclude_id)
    if conn.execute(query, params).fetchone():
        raise HTTPException(status_code=409, detail="You already have a booking overlapping that time slot")


def _check_hall_slot_conflict(conn, date: str, start_time: str, end_time: str, exclude_id: int | None = None):
    """Raise 409 if any approved booking from any user overlaps this slot."""
    query = """
        SELECT id FROM bookings
        WHERE date = ? AND status = 'approved'
        AND start_time < ? AND end_time > ?
    """
    params = [date, end_time, start_time]
    if exclude_id is not None:
        query += " AND id != ?"
        params.append(exclude_id)
    if conn.execute(query, params).fetchone():
        raise HTTPException(status_code=409, detail="The hall is already booked for that time slot")

router = APIRouter(prefix="/api/bookings", tags=["bookings"])


class BookingRequest(BaseModel):
    date: str
    start_time: str
    end_time: str
    purpose: str
    attendees: int


def _booking_row_to_dict(row) -> dict:
    return {
        "id": row["id"],
        "userId": row["user_id"],
        "userName": row["name"],
        "userEmail": row["email"],
        "date": row["date"],
        "startTime": row["start_time"],
        "endTime": row["end_time"],
        "purpose": row["purpose"],
        "attendees": row["attendees"],
        "status": row["status"],
        "createdAt": row["created_at"],
    }


@router.get("")
def list_my_bookings(user: dict = Depends(get_current_user)):
    with get_conn() as conn:
        rows = conn.execute(
            """SELECT b.*, u.name, u.email FROM bookings b
               JOIN users u ON u.id = b.user_id
               WHERE b.user_id = ? ORDER BY b.date DESC""",
            (user["id"],),
        ).fetchall()
    return [_booking_row_to_dict(r) for r in rows]


@router.post("", status_code=status.HTTP_201_CREATED)
def create_booking(body: BookingRequest, user: dict = Depends(get_current_user)):
    _validate_booking_fields(body)
    with get_conn() as conn:
        if conn.execute("SELECT 1 FROM blocked_dates WHERE date = ?", (body.date,)).fetchone():
            raise HTTPException(status_code=409, detail="Date is blocked")
        _check_hall_slot_conflict(conn, body.date, body.start_time, body.end_time)
        _check_user_slot_conflict(conn, user["id"], body.date, body.start_time, body.end_time)
        cursor = conn.execute(
            """INSERT INTO bookings (user_id, date, start_time, end_time, purpose, attendees, status, created_at)
               VALUES (?, ?, ?, ?, ?, ?, 'pending', ?)""",
            (user["id"], body.date, body.start_time, body.end_time, body.purpose, body.attendees,
             date_type.today().isoformat()),
        )
        booking_id = cursor.lastrowid
    return {"id": booking_id, "status": "pending"}


@router.delete("/{booking_id}", status_code=status.HTTP_204_NO_CONTENT)
def cancel_booking(booking_id: int, user: dict = Depends(get_current_user)):
    """Cancel a pending booking that belongs to the current user."""
    with get_conn() as conn:
        row = conn.execute(
            "SELECT id, user_id, status FROM bookings WHERE id = ?", (booking_id,)
        ).fetchone()
        if not row:
            raise HTTPException(status_code=404, detail="Booking not found")
        if row["user_id"] != user["id"]:
            raise HTTPException(status_code=403, detail="Not your booking")
        if row["status"] != "pending":
            raise HTTPException(status_code=409, detail="Only pending bookings can be cancelled")
        conn.execute("DELETE FROM bookings WHERE id = ?", (booking_id,))


@router.put("/{booking_id}")
def update_booking(booking_id: int, body: BookingRequest, user: dict = Depends(get_current_user)):
    """Update a pending booking that belongs to the current user."""
    _validate_booking_fields(body)
    with get_conn() as conn:
        row = conn.execute(
            "SELECT id, user_id, status FROM bookings WHERE id = ?", (booking_id,)
        ).fetchone()
        if not row:
            raise HTTPException(status_code=404, detail="Booking not found")
        if row["user_id"] != user["id"]:
            raise HTTPException(status_code=403, detail="Not your booking")
        if row["status"] != "pending":
            raise HTTPException(status_code=409, detail="Only pending bookings can be modified")
        if conn.execute("SELECT 1 FROM blocked_dates WHERE date = ?", (body.date,)).fetchone():
            raise HTTPException(status_code=409, detail="Date is blocked")
        _check_hall_slot_conflict(conn, body.date, body.start_time, body.end_time, exclude_id=booking_id)
        _check_user_slot_conflict(conn, user["id"], body.date, body.start_time, body.end_time, exclude_id=booking_id)
        conn.execute(
            "UPDATE bookings SET date=?, start_time=?, end_time=?, purpose=?, attendees=? WHERE id=?",
            (body.date, body.start_time, body.end_time, body.purpose, body.attendees, booking_id),
        )
    return {"id": booking_id, "status": "pending"}


@router.get("/availability")
def get_availability(user: dict = Depends(get_current_user)):
    """Returns approved booking dates and blocked dates for calendar display."""
    with get_conn() as conn:
        approved = [r["date"] for r in conn.execute(
            "SELECT date FROM bookings WHERE status = 'approved'"
        ).fetchall()]
        blocked = [{"date": r["date"], "reason": r["reason"]}
                   for r in conn.execute("SELECT date, reason FROM blocked_dates").fetchall()]
    return {"approvedDates": approved, "blockedDates": blocked}

File: frontend/convex/bookings.ts
import { mutation, query, internalQuery } from "./_generated/server";
import { v, ConvexError } from "convex/values";
import { Id } from "./_generated/dataModel";
import { MutationCtx, QueryCtx } from "./_generated/server";

const HALL_OPEN = "06:00";
const HALL_CLOSE = "20:00";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

async function getAuthUser(ctx: QueryCtx | MutationCtx) {
  const identity = await ctx.auth.getUserIdentity();
  if (!identity) throw new ConvexError("Not authenticated");
  const userId = identity.subject.split("|")[0] as Id<"users">;
  const user = await ctx.db.get(userId);
  if (!user) throw new ConvexError("User not found");
  return user;
}

function validateFields(args: {
  date: string;
  startTime: string;
  endTime: string;
  attendees: number;
}) {
  const today = new Date().toISOString().slice(0, 10);
  if (args.date < today) throw new ConvexError("Cannot book a date in the past");
  if (args.startTime < HALL_OPEN || args.startTime >= HALL_CLOSE)
    throw new ConvexError(`Start time must be between ${HALL_OPEN} and ${HALL_CLOSE}`);
  if (args.endTime > HALL_CLOSE)
    throw new ConvexError(`End time cannot be after ${HALL_CLOSE}`);
  if (args.endTime <= args.startTime)
    throw new ConvexError("End time must be after start time");
  if (args.attendees < 1 || args.attendees > 50)
    throw new ConvexError("Attendees must be between 1 and 50");
}

async function checkHallConflict(
  ctx: MutationCtx,
  date: string,
  startTime: string,
  endTime: string,
  excludeId?: Id<"bookings">
) {
  const approved = await ctx.db
    .query("bookings")
    .withIndex("by_date_and_status", (q) =>
      q.eq("date", date).eq("status", "approved")
    )
    .take(100);

  const conflict = approved.find(
    (b) =>
      b._id !== excludeId &&
      b.startTime < endTime &&
      b.endTime > startTime
  );
  if (conflict) throw new ConvexError("The hall is already booked for that time slot");
}

async function checkUserConflict(
  ctx: MutationCtx,
  userId: Id<"users">,
  date: string,
  startTime: string,
  endTime: string,
  excludeId?: Id<"bookings">
) {
  const userBookings = await ctx.db
    .query("bookings")
    .withIndex("by_userId_and_date", (q) =>
      q.eq("userId", userId).eq("date", date)
    )
    .take(50);

  const conflict = userBookings.find(
    (b) =>
      b._id !== excludeId &&
      (b.status === "pending" || b.status === "approved") &&
      b.startTime < endTime &&
      b.endTime > startTime
  );
  if (conflict)
    throw new ConvexError("You already have a booking overlapping that time slot");
}

// ---------------------------------------------------------------------------
// Public queries
// ---------------------------------------------------------------------------

/** Returns the current user's bookings, newest first. */
export const getMyBookings = query({
  args: {},
  handler: async (ctx) => {
    const user = await getAuthUser(ctx);
    const bookings = await ctx.db
      .query("bookings")
      .withIndex("by_userId", (q) => q.eq("userId", user._id))
      .order("desc")
      .take(100);
    return bookings.map((b) => ({
      id: b._id,
      userId: b.userId,
      userName: user.name ?? "",
      userEmail: user.email ?? "",
      date: b.date,
      startTime: b.startTime,
      endTime: b.endTime,
      purpose: b.purpose,
      attendees: b.attendees,
      status: b.status,
      createdAt: b._creationTime,
    }));
  },
});

/** Returns approved booking dates and blocked dates for the calendar. */
export const getAvailability = query({
  args: {},
  handler: async (ctx) => {
    const identity = await ctx.auth.getUserIdentity();
    if (!identity) throw new ConvexError("Not authenticated");

    const approved = await ctx.db
      .query("bookings")
      .withIndex("by_status", (q) => q.eq("status", "approved"))
      .take(500);

    const blocked = await ctx.db
      .query("blockedDates")
      .order("asc")
      .take(200);

    return {
      approvedDates: approved.map((b) => b.date),
      blockedDates: blocked.map((b) => ({ date: b.date, reason: b.reason })),
    };
  },
});

// ---------------------------------------------------------------------------
// Public mutations
// ---------------------------------------------------------------------------

const bookingArgs = {
  date: v.string(),
  startTime: v.string(),
  endTime: v.string(),
  purpose: v.string(),
  attendees: v.number(),
};

/** Create a new booking request. */
export const createBooking = mutation({
  args: bookingArgs,
  handler: async (ctx, args) => {
    const user = await getAuthUser(ctx);
    validateFields(args);

    const blocked = await ctx.db
      .query("blockedDates")
      .withIndex("by_date", (q) => q.eq("date", args.date))
      .unique();
    if (blocked) throw new ConvexError("Date is blocked");

    await checkHallConflict(ctx, args.date, args.startTime, args.endTime);
    await checkUserConflict(ctx, user._id, args.date, args.startTime, args.endTime);

    const id = await ctx.db.insert("bookings", {
      userId: user._id,
      date: args.date,
      startTime: args.startTime,
      endTime: args.endTime,
      purpose: args.purpose,
      attendees: args.attendees,
      status: "pending",
    });
    return { id, status: "pending" };
  },
});

/** Cancel a pending booking that belongs to the current user. */
export const cancelBooking = mutation({
  args: { bookingId: v.id("bookings") },
  handler: async (ctx, args) => {
    const user = await getAuthUser(ctx);
    const booking = await ctx.db.get(args.bookingId);
    if (!booking) throw new ConvexError("Booking not found");
    if (booking.userId !== user._id) throw new ConvexError("Not your booking");
    if (booking.status !== "pending")
      throw new ConvexError("Only pending bookings can be cancelled");
    await ctx.db.delete(args.bookingId);
  },
});

/** Update a pending booking that belongs to the current user. */
export const updateBooking = mutation({
  args: { bookingId: v.id("bookings"), ...bookingArgs },
  handler: async (ctx, args) => {
    const user = await getAuthUser(ctx);
    const booking = await ctx.db.get(args.bookingId);
    if (!booking) throw new ConvexError("Booking not found");
    if (booking.userId !== user._id) throw new ConvexError("Not your booking");
    if (booking.status !== "pending")
      throw new ConvexError("Only pending bookings can be modified");

    validateFields(args);

    const blocked = await ctx.db
      .query("blockedDates")
      .withIndex("by_date", (q) => q.eq("date", args.date))
      .unique();
    if (blocked) throw new ConvexError("Date is blocked");

    await checkHallConflict(ctx, args.date, args.startTime, args.endTime, args.bookingId);
    await checkUserConflict(ctx, user._id, args.date, args.startTime, args.endTime, args.bookingId);

    await ctx.db.patch(args.bookingId, {
      date: args.date,
      startTime: args.startTime,
      endTime: args.endTime,
      purpose: args.purpose,
      attendees: args.attendees,
    });
    return { id: args.bookingId, status: "pending" };
  },
});

// ---------------------------------------------------------------------------
// Internal queries (used by chat action)
// ---------------------------------------------------------------------------

/** Returns availability context string for the AI system prompt. */
export const getAvailabilityContext = internalQuery({
  args: {},
  handler: async (ctx) => {
    const approved = await ctx.db
      .query("bookings")
      .withIndex("by_status", (q) => q.eq("status", "approved"))
      .take(500);
    const blocked = await ctx.db.query("blockedDates").order("asc").take(200);

    const lines: string[] = [];
    if (approved.length > 0)
      lines.push(`Hall-wide approved booking dates: ${approved.map((b) => b.date).join(", ")}`);
    if (blocked.length > 0)
      lines.push(`Blocked dates: ${blocked.map((b) => `${b.date} (${b.reason})`).join(", ")}`);
    if (lines.length === 0)
      lines.push("No dates are currently booked or blocked.");
    return lines.join("\n");
  },
});

/** Returns a user's bookings as a string for the AI system prompt. */
export const getUserBookingsContext = internalQuery({
  args: { userId: v.id("users") },
  handler: async (ctx, args) => {
    const bookings = await ctx.db
      .query("bookings")
      .withIndex("by_userId", (q) => q.eq("userId", args.userId))
      .order("asc")
      .take(50);
    if (bookings.length === 0) return "This user has no bookings yet.";
    return bookings
      .map(
        (b) =>
          `- [booking_id=${b._id}] ${b.date} ${b.startTime}–${b.endTime}: ${b.purpose.slice(0, 200).replace(/\n/g, " ")} (${b.attendees} attendees) — status: ${b.status}`
      )
      .join("\n");
  },
});
